File: src/main/java/testbed/app/AppController.java
package testbed.app;

import javafx.application.Platform;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.scene.chart.LineChart;
import javafx.scene.chart.NumberAxis;
import javafx.scene.chart.XYChart;
import javafx.scene.control.Button;
import javafx.scene.control.TextField;
import javafx.scene.layout.VBox;
import testbed.common.DefaultRandomGenerator;
import testbed.core.AlphaTestbed;
import testbed.core.EvaluationProtocol;
import testbed.core.MetricSubscriber;
import testbed.interfaces.Metric;
import testbed.interfaces.Scenario;
import testbed.interfaces.TrustModel;
import testbed.metric.CumulativeNormalizedUtility;
import testbed.metric.DefaultOpinionCost;
import testbed.metric.KendallsTauA;
import testbed.scenario.TransitiveOpinionProviderSelection;
import testbed.trustmodel.SimpleSelectingOpinionProviders;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class AppController implements MetricSubscriber {

    public TextField input;
    public Button start;
    public Button init;
    public Button stop;
    public VBox vbox;

    private final NumberAxis xAxis = new NumberAxis();
    private final NumberAxis yAxis = new NumberAxis();
    private final LineChart<Number, Number> chart = new LineChart<>(xAxis, yAxis);

    private Metric accuracy, utility, opinionCost;
    private XYChart.Series<Number, Number> accuracySeries, utilitySeries, opinionCostSeries;

    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    @FXML
    protected void initialize() {
        vbox.getChildren().add(chart);
    }

    public void initButton(ActionEvent event) {
        // series.getData().clear();
        accuracy = new KendallsTauA();
        accuracySeries = new XYChart.Series<>();
        accuracySeries.setName(accuracy.toString());
        chart.getData().add(accuracySeries);

        utility = new CumulativeNormalizedUtility();
        utilitySeries = new XYChart.Series<>();
        utilitySeries.setName(utility.toString());
        chart.getData().add(utilitySeries);

        opinionCost = new DefaultOpinionCost();
        opinionCostSeries = new XYChart.Series<>();
        opinionCostSeries.setName(opinionCost.toString());
        chart.getData().add(opinionCostSeries);
    }


    public void startButton(ActionEvent event) {
        executor.execute(() -> {
            final TrustModel<?> model = new SimpleSelectingOpinionProviders();
            model.setRandomGenerator(new DefaultRandomGenerator(0));
            model.initialize();

            final Scenario scenario = new TransitiveOpinionProviderSelection();
            scenario.setRandomGenerator(new DefaultRandomGenerator(0));
            scenario.initialize(100, 0.05, 0.1, 1d, 1d);

            final Map<Metric, Object[]> metrics = new HashMap<>();
            metrics.put(accuracy, null);
            metrics.put(utility, null);
            metrics.put(opinionCost, null);

            final EvaluationProtocol ep = AlphaTestbed.getProtocol(model, scenario, metrics);
            ep.subscribe(this);

            for (int time = 1; time <= 500; time++) {
                ep.step(time);
            }
        });
    }

    public void stopButton(ActionEvent event) {
        executor.shutdownNow();
    }

    @Override
    public void update(final EvaluationProtocol instance) {
        Platform.runLater(() -> {
            accuracySeries.getData().add(new XYChart.Data<>(instance.getTime(), instance.getResult(0, accuracy)));
            utilitySeries.getData().add(new XYChart.Data<>(instance.getTime(), instance.getResult(0, utility)));
            opinionCostSeries.getData().add(new XYChart.Data<>(instance.getTime(), instance.getResult(0, opinionCost)));
        });
    }
}

File: src/main/kotlin/testbed/app/AppController.kt
package testbed.app

import javafx.application.Platform
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.scene.chart.LineChart
import javafx.scene.chart.NumberAxis
import javafx.scene.chart.XYChart
import javafx.scene.control.Button
import javafx.scene.control.TextField
import javafx.scene.layout.VBox
import testbed.common.DefaultRandomGenerator
import testbed.core.AlphaTestbed
import testbed.core.EvaluationProtocol
import testbed.core.MetricSubscriber
import testbed.gui.ParametersGUI
import testbed.interfaces.*
import java.util.concurrent.Executors

class AppController : MetricSubscriber {
    @FXML
    lateinit var input: TextField
    @FXML
    lateinit var init: Button
    @FXML
    lateinit var stop: Button
    @FXML
    lateinit var verticalBox: VBox

    private val chart = LineChart(
            NumberAxis().apply {
                tickUnit = 25.0
                lowerBound = 0.0
            },
            NumberAxis().apply {
                upperBound = 1.0
                lowerBound = 0.0
                tickUnit = 0.05
                isAutoRanging = false
            })

    private val metricData = HashMap<Metric, XYChart.Series<Number, Number>>()

    private val executor = Executors.newSingleThreadExecutor()

    @FXML
    fun initialize() {
        verticalBox.apply {
            children += chart
        }
        input.text = "1"
    }

    fun initButton(event: ActionEvent) {
        val gui = ParametersGUI(AppController::class.java.classLoader)
        gui.setBatchRun(true)

        val answer = gui.showDialog()

        if (answer != 0) {
            return
        }

        val seed = Integer.parseInt(input.text)

        val scenario = gui.setupParameters[0] as Scenario
        scenario.setRandomGenerator(DefaultRandomGenerator(seed))
        scenario.initialize(*gui.scenarioParameters)

        val trustModel = gui.setupParameters[1] as TrustModel<*>
        trustModel.setRandomGenerator(DefaultRandomGenerator(seed))
        trustModel.initialize(*gui.trustModelParameters)

        val metrics = HashMap<Metric, Array<Any>>()
        gui.setupParameters[2]?.let {
            metrics[it as Accuracy] = gui.accuracyParameters
            metricData[it] = XYChart.Series<Number, Number>().apply { name = it.toString() }
            chart.data.add(metricData[it])
        }
        gui.setupParameters[3]?.let {
            metrics[it as Utility] = gui.utilityParameters
            metricData[it] = XYChart.Series<Number, Number>().apply { name = it.toString() }
            chart.data.add(metricData[it])
        }
        gui.setupParameters[4]?.let {
            metrics[it as OpinionCost] = gui.opinionCostParameters
            metricData[it] = XYChart.Series<Number, Number>().apply { name = it.toString() }
            chart.data.add(metricData[it])
        }

        val duration = gui.setupParameters[5] as Int

        val protocol = AlphaTestbed.getProtocol(trustModel, scenario, metrics)
        protocol.subscribe(this)

        executor.execute {
            for (time in 1..duration) {
                protocol.step(time)
            }
        }
    }

    fun stopButton(event: ActionEvent) {
        chart.data.clear()
        metricData.clear()
    }

    override fun update(instance: EvaluationProtocol) {
        Platform.runLater {
            for ((metric, series) in metricData) {
                series.data.add(XYChart.Data(instance.time, instance.getResult(0, metric)))
            }
        }
    }
}